### DISCUSSION TITLE
Improving performance on Raspberry PI

### DISCUSSION THREAD

**Discussion by gitbls** [strongswan/strongswan#1345] [Q&A]
Until recently I haven't bothered to check the performance of my Pi4-based VPN. I finally found some time to look into this, and discovered that iperf3 over the VPN (Pi4-RasPiOS to Pi4-RasPiOS) using `proposals=aes128-sha256-x25519` and `esp_proposals = aes128gcm128-x25519` was **186Mb/sec**.

With `proposals = chacha20poly1305-sha512-curve25519` and `esp_proposals = chacha20poly1305-sha512` I'm seeing **332Mb/sec**.

I'm not expecting to miraculously see near-line speed performance, but wondering if anyone has any guidance on what proposals/esp_proposals to use that might offer better performance than what I'm seeing with chacha20poly1305.

Finally, since we can't change the ciphers used by iOS and Windows, what proposals/esp_proposals might be supported by those OS that would improve performance?

Don't get me wrong, I don't think the performance is too bad, but it is a bit challenging when compared to wg, which clocks in at **885Mb/sec** on the exact same configuration.

Thanks!

**Comment by jxcang** [comment]
332 Mb/s is impressive. When I run a simple bridge between eth0 and USB eth1, I can only get 680 Mb/s before the irq goes to 100.
You can try to use a less CPU intensive algorithm, but at the cost of security.

You can change the algos on Windows through RegEdit. iOS proposes 5 or 7 algos at once for you to choose from, I believe (from: I spent hours with Cisco smartNet to get iOS working with their RV product line). I can go look for WireShark captures of iOS phase 1.

What is wg? I could not even get another RPi 4 now. My current RPi 4 is receiving 10 HD surveillance camera streams.
I am also looking for RPi alternatives for both better pricing, and have more than one Ethernet jacks on the board. NanoPi R2S and some Banana Pis seemed promising.

**Comment by gitbls** [comment]
Thanks for the feedback. I agree that 332Mb/s isn't bad, but next to Wireguard (aka wg) there's room for improvement. I'm not suggesting that strongSwan needs to achieve near-line speed as wg does, just looking for possible paths to improve the performance of strongSwan. At the moment, strongSwan still has advantages over wg in many scenarios.

I would definitely like to try a less CPU intenstive algorithm, but I haven't found any resources that provide indications of CPU utilization even as a starting point. 

Short of setting up a test jig to figure this out myself, are there any resources that might help sort this out?

**Reply by jxcang** [reply]
You can search for encryption algorithm vs performance in terms of completion time or CPU load. This comparison is not really scientific and should always be taken with a grain of salt in deference to empirical evidence. There are comparisons made by somebody. Please excuse my incompleteness in attribution. https://www.cs.wustl.edu/~jain/cse567-06/ftp/encryption_perf/

I was googling Wireguard and the Wikipedia says [partially paraphrased] it is designed to be better performance than IPsec. So this is a foregone conclusion?

I like IPsec and strongswan because it is what is used on Cisco RV series router and I have one at home. Cisco RV also has OpenVPN which I haven't learned it. I was studying Linux operating system, then became interested in Linux networking, then got further digressed into ip xfrm and ipsec.

**Comment by gitbls** [comment]
Thanks for your suggestions! I haven't found anything that looks faster than chacha20poly1305-sha512-curve25519, so maybe that's as good as it gets. For now, at least.
